agent/src/flow_generator/protocol_logs/mq/kafka.rs

```diff
@@ -352,22 +452,125 @@ impl KafkaLog {
         if proto != IpProtocol::TCP {
             return Err(Error::InvalidIpProtocol);
         }
-        if payload.len() < KAFKA_REQ_HEADER_LEN {
-            return Err(Error::KafkaLogParseFailed);
-        }
+
         match direction {
             PacketDirection::ClientToServer => {
+                if payload.len() < KAFKA_REQ_HEADER_LEN {
+                    return Err(Error::KafkaLogParseFailed);
+                }
                 self.request(payload, false, info)?;
                 self.perf_stats.as_mut().map(|p| p.inc_req());
             }
             PacketDirection::ServerToClient => {
+                if payload.len() < KAFKA_RESP_HEADER_LEN {
+                    return Err(Error::KafkaLogParseFailed);
+                }
                 self.response(payload, info)?;
                 self.perf_stats.as_mut().map(|p| p.inc_resp());
             }
         }
         Ok(())
     }
 
+    fn get_topics_name(
+        &mut self,
+        api_key: u16,
+        api_version: u16,
+        payload: &[u8],
+    ) -> Option<String> {
+        let mut _fixed_offset = kafka_apiversion_topic_fixed_offset!(api_key, api_version);
+        if _fixed_offset == usize::max_value() {
+            return None;
+        }
+        if _fixed_offset > payload.len() {
+            return None;
+        }
+        match api_key {
+            KAFKA_PRODUCE => {
+                if api_version >= 3 && api_version <= 8 {
+                    let tid_len = read_i16_be(&payload[0..2]);
+                    if tid_len > 0 {
+                        _fixed_offset = _fixed_offset + tid_len as usize
+                    }
+                }
+                // Significant improvements since version 9
+                if api_version == 9 {
+                    let tid_len = payload[0];
+                    if tid_len > 0 {
+                        _fixed_offset = _fixed_offset + tid_len as usize
+                    }
+                    if _fixed_offset + 1 + payload[_fixed_offset] as usize > payload.len() {
+                        return None;
+                    }
+                    let (_, _, result) = UTF_16BE.decode(
```

Comment:
What is using `UTF_16BE.decode` and `String::from_utf8_lossy` after it for?

Reply:
I don't have a good method here. I tried some decoding methods. In my environment, I barely got a decoder. I used `UTF_16BE.decode` to make a judgment and read it normally with `String::from_utf8_lossy`. Because of compatibility, I haven't found the exact reason in my production environment. I added a todo here to wait for future repair.

Reply:
This indeed requires more inspection. According to encoding-rs [doc](https://docs.rs/encoding_rs/latest/encoding_rs/struct.Encoding.html#method.decode), the `result` that `UTF_16BE.decode` returns indicates whether there were "malformed sequences" in the buffer, which I suppose should be the bad case? The content of payload should be examined. If they are UTF-16, they should not be decoded as UTF-8.